### DISCUSSION TITLE
How does multipart processing work on the server when using the Fetch API?

### DISCUSSION THREAD

**Discussion by ghost** [denoland/deno#13487] [Q&A]
I'm writing a web server in Deno, and I'm at the stage where I need to do multipart processing.

Previously when I did this in Node, I needed to specify and handle configuration settings like where file uploads go and how much memory can be used during multipart processing. However, based on the information I can gather from the docs, there's no need to specify these settings for multipart thanks to `await req.formData()`, which takes care of the low-level stuff for you.

But now I'm curious about the specifics of multipart processing when using `await req.formData()` on the server. (i.e. How much memory does this use? Where do files go? Do they stay there or should I delete/move them?) My guess is that I don't need to worry about those things because Deno usually uses best practices by default, but I'm having a hard time checking that hunch.

I'd love some pointers on this if possible. I've searched for a bit but came up empty handed, and I felt a discussion question might be helpful for others as well.

**Comment by lucacasonato** [comment]
For FormData, everything is currently stored in memory. In the future individual file blocks may be stored on disk. The amount of data stored will be capped just like in browsers, and exceeding those limits will cause formData to return a QuotaExceeded error.

It does seem reasonable that one may want to cap the total received length. Currently this can be done by intercepting the body stream and capping it in length. It would be much nicer if this could be done like this `req.formData({ maxByteLength: 4000 })`.

If that last option seems reasonable, I'll bring it up with the WHATWG for inclusion in the spec.

**Reply by lucacasonato** [reply]
Problem with `Content-Length` is that it is not reliable on HTTP/2. For HTTP/2, the content length header may be different to the actual length of the header.

**Reply by andreubotella** [reply]
I think something more useful might be an option to ignore uploaded files and only retrieve textual form fields, or to only retrieve files with specific form names, perhaps downloading them into a folder in the filesystem. That seems like an API that would be useful server-side, but that would not make too much sense client-side.

Of course, for Deno, `req.formData()` couldn't automatically download files into the filesystem before something like #10538 is implemented.
